server/html.go

```diff
@@ -131,34 +137,79 @@ func (s *Server) Index(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
-// NewInbox creates a new inbox and returns details to the user
-func (s *Server) NewInbox(w http.ResponseWriter, r *http.Request) {
+// NewRandomInbox generates a new Inbox with a random route and host from availabile options.
+func (s *Server) NewRandomInbox(w http.ResponseWriter, r *http.Request) {
 	i := data.NewInbox()
-	sess, ok := r.Context().Value(sessionCTXKey).(*sessions.Session)
-	if !ok {
-		log.Printf("New Inbox: failed to get sess var. Sess not of type sessions.Session actual type: %v", reflect.TypeOf(sess))
+	i.Address = s.eg.NewRandom()
+
+	exist, err := s.db.EmailAddressExists(i.Address) // while it's VERY unlikely that the email address already exists but lets check anyway
+	if err != nil {
+		log.Printf("NewRandomInbox: failed to check if email exists: %v", err)
 		http.Error(w, "Failed to generate email", http.StatusInternalServerError)
 		return
 	}
 
-	i.Address = s.eg.NewRandom()
+	if exist {
+		log.Printf("NewRandomInbox: email already exisists: %v", i.Address)
+		http.Error(w, "Failed to generate email", http.StatusInternalServerError)
+		return
+	}
 
-	exist, err := s.db.EmailAddressExists(i.Address) // while it's VERY unlikely that the email address already exists but lets check anyway
+	s.CreateRouteFromInbox(w, r, i)
+}
+
+// NewNamedInbox generates a new Inbox with a specific route and host.
+func (s *Server) NewNamedInbox(w http.ResponseWriter, r *http.Request) {
+	route := r.PostFormValue("route")
+	host := r.PostFormValue("host")
+	if route == "" {
+		log.Printf("NewNamedInbox: missing required `route` parameter")
+		http.Error(w, "Failed to generate email", http.StatusInternalServerError)
+		return
+	} else if host == "" {
+		log.Printf("NewNamedInbox: missing required `host` parameter")
+		http.Error(w, "Failed to generate email", http.StatusInternalServerError)
+		return
+	}
+
+	address, err := s.eg.NewFromRouteAndHost(route, host)
+	if err != nil {
+		log.Printf("NewNamedInbox: failed to create new inbox address: %v", err)
+		http.Error(w, "Failed to generate email", http.StatusInternalServerError)
+		return
+	}
+
+	i := data.NewInbox()
+	i.Address = address
+
+	exist, err := s.db.EmailAddressExists(i.Address)
 	if err != nil {
-		log.Printf("New Inbox: failed to check if email exists: %v", err)
+		log.Printf("NewNamedInbox: failed to check if email exists: %v", err)
 		http.Error(w, "Failed to generate email", http.StatusInternalServerError)
 		return
 	}
 
 	if exist {
-		log.Printf("NewInbox: email already exisists: %v", err)
+		log.Printf("NewNamedInbox: email already exisists: %v", i.Address)
+		http.Error(w, "Failed to generate email", http.StatusInternalServerError)
```

Comment:
I think this error should be a bit more descriptive. Let’s tell the user that the email is taken. Unless you can think of some reason not to.